[Term Entry] PyTorch Tensor Operations .index_reduce()

mamtawardhani opened the pull request:
<!--
👋 Hi, thanks for submitting a PR to Codecademy Docs! 💖
Please fill out all fields below to ensure your PR is reviewed quickly.

**IMPORTANT**

If you would like to receive credit for your contribution to an entry, make sure to have your Codecademy user profile linked to your GitHub account:

1. Go to your Codecademy dashboard.
2. Click on your profile image in the top-right and then choose "Profile".
3. Then click "Edit Profile".
4. In the "GitHub Username" field, add your username (without the @).
5. Click "Save Changes"

Of course, you can opt not to do this and be listed as an "Anonymous contributor", instead. :)
-->

### Description
Added a new Term Entry for PyTorch Tensor Operations `.index_reduce()`

### Issue Solved
Closes #5616 

### Type of Change

<!--- Please delete or cross off the bullet point(s) that are irrelevant to this PR: -->

- Adding a new entry

### Checklist

<!-- Please check ALL the boxes: -->

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [ ] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] I have linked any issues that are relevant to this PR in the `Issues Solved` section.

<!--
Having trouble with the PR checker? Here are some common issues and resolutions:

- verify_formatting is failing
  - run `yarn format path/to/markdown/file.md` or `yarn format:all` and commit the results
- verify_lint is failing
  - same as above
  - if verify_lint is still failing, running `yarn lint` locally should let you know what needs to be changed by hand
- test is failing
  - ensure any new markdown files have a `Title` and `Description` defined in their metadata
  - ensure any new markdown files only contain alphanumerics and dashes in their file names and have the same name as their parent directory
  - if that looks ok, running `yarn test` locally should let you know what the issue is
-->

PragatiVerma18 commented on content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md:
````diff
@@ -0,0 +1,64 @@
+Tensor.index_reduce_(dim, index, source, reduce, *, include_self=True)
+```
+
+- `dim`: The dimension along which the reduction is performed. It specifies which axis of the tensor is being reduced.
````
There is repetition in this explanation. We can rewrite it simply as follows:

```suggestion
- `dim`: The axis of the tensor along which the reduction is performed.
```

PragatiVerma18 commented on content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md:
```diff
@@ -0,0 +1,64 @@
+# Indices mapping source to target
+index = torch.tensor([0, 1, 0, 1])
+
+# Perform in-place reduction using 'mean'
```
The parameter `dim=0` is used in the example, which is fine, but it may not be immediately clear why dimension 0 was chosen. For users unfamiliar with PyTorch tensor dimensions, a brief explanation would help.

```suggestion
# Perform in-place reduction using 'mean' along the 0th dimension (rows)
```

PragatiVerma18 commented on content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md:
```diff
@@ -0,0 +1,64 @@
+  - 'paths/computer-science'
+---
+
+In PyTorch, **`.index_reduce()`** performs a reduction operation (such as sum, prod, or mean) on a [tensor](https://www.codecademy.com/resources/docs/pytorch/tensors) along a specified dimension, using an index tensor to map input elements to positions in the output tensor, effectively aggregating values with the same index.
```
We have not explained why we are using `.index_reduce()` in some places and why `.index_reduce_()` in others; this is confusing. Please add an explanation somewhere.

PragatiVerma18 commented on content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md:
```diff
@@ -0,0 +1,64 @@
+source = torch.tensor([1.0, 2.0, 3.0, 4.0])
+
+# Indices mapping source to target
+index = torch.tensor([0, 1, 0, 1])
```
It's important to highlight the type requirement for the `index` tensor, as PyTorch typically requires integer types for indexing. The operation may fail if the `index` tensor is not of an integer type.

```suggestion
index = torch.tensor([0, 1, 0, 1], dtype=torch.long)  # Ensure index tensor is of type 'long'
```

PragatiVerma18 commented on content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md:
````diff
@@ -0,0 +1,64 @@
+```py
+import torch
+
+# Define the target tensor
+target = torch.zeros(5)
````
The input tensors (`target`, `source`, and `index`) have different shapes. The `index` tensor has 4 elements, and the `source` tensor has 4 elements, but `target` is initialized with 5 elements. 

This may cause confusion because the `target` tensor should match the number of output locations determined by the `index` tensor.

Ensure that the `target` tensor matches the number of unique indices in `index`. If `index` contains values mapping to positions 0 and 1, for example, the `target` tensor should have at least 2 elements.

```suggestion
# Define the target tensor
target = torch.zeros(2)
```

Please update output accordingly.

PragatiVerma18 reviewed (changes requested):
Hey @mamtawardhani, thanks for adding this entry. I have some suggestions, please check and do the required fixes. Let me know if you have any questions. Happy contributing! 🥳

mamtawardhani commented:
Hi @PragatiVerma18, thanks for the feedback - I've made the changes, also I've updated the entry to be `.index_reduce_()` instead of `.index_reduce(), as we are working on tensor operations. I've updated the issue as well

PragatiVerma18 reviewed (approved):
Hey @mamtawardhani, thanks for doing the changes so quickly. This looks good to be merged! 🚀
